File: backend/app/models/task.py
import uuid
from datetime import datetime, timezone
from enum import Enum as PyEnum
from sqlalchemy import Column, String, Boolean, DateTime, Date, Time, ForeignKey, Enum, Text
from sqlalchemy.dialects.sqlite import CHAR
from sqlalchemy.orm import relationship
from app.database import Base


class Priority(str, PyEnum):
    high = "high"
    medium = "medium"
    low = "low"


class Status(str, PyEnum):
    pending = "pending"
    completed = "completed"
    cancelled = "cancelled"


class TimeHorizon(str, PyEnum):
    today = "today"
    week = "week"
    someday = "someday"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(CHAR(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(CHAR(36), ForeignKey("users.id"), nullable=False)
    goal_id = Column(CHAR(36), ForeignKey("goals.id"), nullable=True)
    title = Column(String(500), nullable=False)
    description = Column(Text, default="")
    agent_notes = Column(Text, nullable=True)
    generated_by_agent = Column(Boolean, default=False)
    status = Column(Enum(Status), default=Status.pending)
    completed_at = Column(DateTime, nullable=True)
    priority = Column(Enum(Priority), default=Priority.medium)
    due_date = Column(Date, nullable=True)
    due_time = Column(Time, nullable=True)
    time_horizon = Column(Enum(TimeHorizon), default=TimeHorizon.today)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="tasks")
    goal = relationship("Goal", back_populates="tasks")
    subtasks = relationship("Subtask", back_populates="task", cascade="all, delete-orphan")

File: frontend/src/app/core/models/task.ts
// Task-related types matching backend schemas

export type Priority = 'high' | 'medium' | 'low';

export type Status = 'pending' | 'completed' | 'cancelled';

export type TimeHorizon = 'today' | 'week' | 'someday';

export type SubtaskStatus = 'pending' | 'completed' | 'cancelled';

export interface Subtask {
  id: string;
  task_id: string;
  title: string;
  description: string;
  agent_notes: string;
  status: SubtaskStatus;
  order: number;
  generated_by_agent: boolean;
  created_at: string;
}

export interface Task {
  id: string;
  user_id: string;
  goal_id: string | null;
  title: string;
  description: string;
  agent_notes: string | null;
  generated_by_agent: boolean;
  status: Status;
  completed_at: string | null;
  priority: Priority;
  due_date: string | null;
  due_time: string | null;
  time_horizon: TimeHorizon;
  created_at: string;
  updated_at: string;
  subtasks: Subtask[];
}

export interface TaskCreate {
  title: string;
  description?: string;
  goal_id?: string | null;
  priority?: Priority;
  due_date?: string | null;
  due_time?: string | null;
  time_horizon?: TimeHorizon;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  goal_id?: string | null;
  priority?: Priority;
  status?: Status;
  due_date?: string | null;
  due_time?: string | null;
  time_horizon?: TimeHorizon;
}
